### PR TITLE
Breaking change in tuples after VS 15.4 update

### PR DESCRIPTION
**Pull request by forki**
With VS Update 15.3.5 we get:

image

With VS Update 15.4 we see:

image

/cc @mexx @drvink

### CODE REVIEW

**Comment by dsyme** [comment]
OK, so this is a regression.  We fixed an issue to translate ``Tuple<int,int>`` into F#'s view of tuple types ``int * int``.  However the F# view of tuple types doesn't support ``.Item1`` etc. properties.

I'm not sure what to do about this, we will need to think about it.

**Comment by forki** [comment]
Tbh I think it should compile in any case since we only use system.tuple everywhere

**Comment by matthid** [comment]
I think whenever the type `System.Tuple` is specified explicitly the compiler would allow `ItemX` otherwise it probably shouldn't?

Besides that `System.Tuple` and `*` should behave exactly the same (and we should be able to change one to the other without semantic changes - with the exception of `ItemX`).

The easier solution would be to allow `ItemX` everywhere, but that might have other side-effects.

**Comment by forki** [comment]
Workaround:

image

**Comment by forki** [comment]
ok this fails in ci_part1 for unrelated reason. I guess master is broken

**Comment by dsyme** [comment]
@forki Yes, there's something fishy with the latest commit "Merge branch 'vs2017-rtm' into master".

**Comment by dsyme** [comment]
> I think whenever the type System.Tuple is specified explicitly the compiler would allow ItemX otherwise it probably shouldn't? Besides that System.Tuple and * should behave exactly the same (and we should be able to change one to the other without semantic changes - with the exception of ItemX). The easier solution would be to allow ItemX everywhere, but that might have other side-effects.

See https://github.com/Microsoft/visualfsharp/issues/3016, https://github.com/Microsoft/visualfsharp/pull/3283

**Comment by matthid** [comment]
> See #3016, #3283

Ok with that in mind another solution would in fact be to not allow users to use `System.Tuple` by either emitting a warning or even an error (because apparently it is hard to write code when you have that anywhere)

**Comment by forki** [comment]
@matthid this code comes from C# interop in real life app. We broke that.

one could argue that's bad design to use tuples in public APIs, but this is still a breaking change

**Comment by matthid** [comment]
@forki Yes non-breaking is only my suggestion above. Problem is as far as I can understand the point of @dsyme is to strictly separate language-tuples from `System.Tuple`. That they are in fact the same is an implementation detail. This is a hard decision to make with what is already released.

Maybe the compiler needs to be less eager in converting to F# tuples as it is right now?

**Comment by forki** [comment]
ok so the test is capturing the bug.

**Comment by curtnichols** [comment]
Found the same issue in our code base, C# calling an F# function that explicitly specifies `System.Tuple` with types. Repro:

```
[<CompiledName("F")>]
let f (p : System.Tuple<bool, int>) =
    printfn "%A %A" p.Item1 p.Item2
```

I'm finding it rather troublesome that one can specify a concrete type and the compiler supersedes the request.

**Comment by dsyme** [comment]
@curtnichols Thank you for reporting this.  Firs,t I want to apologize that this change has come into F# 4 .1 midway instead of being done as a language addition/change in F# 4.2.   The rising use of System.Tuple meant that the #3016 was becoming a more serious issue that would have meant very inconsistent views of the use of the type when encountered via F# source and C# source.  However I hadn't appreciated fully the side effects that addressing the issue would have.  Certainly, an RFC should have been written with full analysis, and it is something I will endeavour to do next week - better late than never.

The change to fix #3016 has effectively made  ``(int * int)`` and ``Tuple<int*int>`` equivalent and regular.  The case you're seeing above is actually a long standing nit in the F# compiler - which we decided long ago not to change - that there is an element of type-directedness in determining the compiled form of a function that accepts tuples as arguments.   Basically a function like this:

    let f1 (p : (bool * int))) = ()

or 

    let f2 ((a,b,c) as d) = ()

will in each case have a compiled form that accepts multiple arguments, rather than one tupled argument.  As you're encountering, it now doesn't matter whether you write ``(bool * int)`` or ``Tuple<bool,int>`` - in either case you now get two arguments.  It's at least now consistent in that the two ways of writing tuple types are equivalent

The original rational for this approach was that the following is quite a common coding pattern:

    let f3 (a,b,c) = ...
    let f4 args = f3 args

In this case, looking only at ``let f4 args = f3 args``, it is reasonable for the programmer to expect that f3 and f4 will have the same compiled form.  (It's intrinsic in the fact there is even a choice about the compiled form that you have to "pay the cost" at some point in the language design - in this case we choice "implicit type-directed de-tupling happens" over "innocuous rebindings change your compiled form".  The F# design has a number of such choices.

The only way I know of to stop the compiler making this choice is now to add a signature file to a module (previously you could use an explicit ``System.Tuple`` as in your code, but that route has now been blocked off).  If you do that, and use

    val f1 : p:(int * bool)  -> unit
    val f2 : p:(int * int * int)  -> unit

then you will get one argument in the compiled form.  If you use

    val f1 : a:int * b:bool  -> unit
    val f2 : a:int * b:int * c:int  -> unit

then you get multiple arguments.  AFAIK there is no way to get this behavior any more without a signature file.  I'm still thinking about what we should do here, and I will try to document the choices in an RFC next week.  We may add such a way, or we could in theory roll back the change that fixed #3016 (though that seems unlikely)

In short the workaround at the moment appears to be to add a signature file.  I understand that's not a great workaround though if your implementation file is large.

**Comment by dsyme** [comment]
@forki How serious is the ``Item1``, ``Item2``... thing for you (or others)?

**Comment by forki** [comment]
For me personally it's no issue at all. I only reported it for @mexx. So he should comment on that.

**Comment by forki** [comment]
I mean since destructuring with (i1, i2) is working it's only an annoying step to go over all the item1 thingies and change the code.

**Comment by forki** [comment]
But then we need to make sure we don't break it ever again.

**Comment by mexx** [comment]
In my case the code base which was broken is not big, I was able to change the code to function again.
For me it's not critical, but I can imagine someone could have more code to change.

**Comment by forki** [comment]
Tbf 8 think it's now pretty usable when you consider that you can basically
destructure all kinds of tuples now with same syntax. Only trouble is that
it is breaking...

Am 15.10.2017 14:52 schrieb "Max Malook" <notifications@github.com>:

In my case the code base which was broken is not big, I was able to change
the code to function again.
For me it's not critical, but I can imagine someone could have more code to
change.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/Microsoft/visualfsharp/pull/3729#issuecomment-336709609>,
or mute the thread

**Comment by curtnichols** [comment]
[Here](https://gist.github.com/curtnichols/7100ccfe528bd5baad83c2ae0e0e1dc4) is a more sinister (but real) repro, simplified. I've attempted the signature file, but as you can see the Sytem.Tuple is not part of a function signature, so my optimism wanes.

The F# side essentially looks like this:

```
module Update =

    open System
    open System.IO

    type FileMapper = Func<string, Tuple<bool, Stream>>

    type UpdateArguments = {
        MapFilePath : FileMapper
    }

    [<CompiledName("F")>]
    let f (args : UpdateArguments) =

        let success, file =     let result = args.MapFilePath.Invoke("name")
                                result.Item1, result.Item2
        printfn "%A %A" success file.CanRead
```

@dsyme 
> The case you're seeing above is actually a long standing nit in the F# compiler....

I think I grok most of what you're saying, but this is obviously an interop issue--which I think you mention as the reason for prioritizing this change. So this was meant to be a breaking change sooner or later? As in, our understanding of how this interop works needs to change?

I write primarily F#, C#, and C++, with interop between the various factions. Few things push my buttons more than compilers not doing what I ask of them.

**Comment by KevinRansom** [comment]
@dsyme 

I made a small modification to @curtnichols' example, to make the surprise of this behavior more clear.

````
    [<CompiledName("F")>]
    let f (args : UpdateArguments) =
        let success, file =     let result:Tuple<bool, Stream> = args.MapFilePath.Invoke("name")
                                result.Item1, result.Item2
        printfn "%A %A" success file.CanRead
````

And it yielded a compile error.
````
>------ Rebuild All started: Project: Library1, Configuration: Debug Any CPU ------
1>C:\Users\kevinr\source\repos\ConsoleApp1\Library1\Library1.fs(17,40): error FS0039: The field, constructor or member 'Item1' is not defined.
1>C:\Users\kevinr\source\repos\ConsoleApp1\Library1\Library1.fs(17,54): error FS0039: The field, constructor or member 'Item2' is not defined.
1>C:\Users\kevinr\source\repos\ConsoleApp1\Library1\Library1.fs(18,33): error FS0072: Lookup on object of indeterminate type based on information prior to this program point. A type annotation may be needed prior to this program point to constrain the type of the object. This may allow the lookup to be resolved.
1>Done building project "Library1.fsproj" -- FAILED.
````
I think this error message would surprise most F# developers. and be pretty hard to explain.

**Comment by varon** [comment]
I could easily be missing something, but wouldn't it be a whole lot simpler and avoid tons of garbage if we could just use the new struct tuples everywhere and get rid of the old class-based tuples?

**Comment by dsyme** [comment]
@varon 

> I could easily be missing something, but wouldn't it be a whole lot simpler and avoid tons of garbage if we could just use the new struct tuples everywhere and get rid of the old class-based tuples?

It's irrelevant to this discussion, and wouldn't solve the issues here

>I think this error message would surprise most F# developers. and be pretty hard to explain.

F# developers don't use ``Tuple`` or ``Item1`` explicitly - they use ``(int * int)`` and pattern matching.  

We could allow the use of ``ItemN`` properties. However these properties are defined in the _compiled_ form of the types, which is relevant for large tuples.  So, for large tuples, would ``(int * ...... * int)`` 10-tuple have a ``Item10`` property?  Or only up to ``Item7`` or ``Item8``?

The correct thing to do in F# is to adjust the code above to this:

    let f (args : UpdateArguments) =
        let success, file = args.MapFilePath.Invoke("name")
        printfn "%A %A" success file.CanRead

As you can see, the code was just a kludge to workaround the fact that F# wasn't understanding that ``Tuple<X,Y>`` is the same as ``(X * Y)``

**Comment by matthid** [comment]
@dsyme I think in an ideal fix for this a variable/symbol declared as `(X * Y)` would not have `ItemN` properties.
But if the developer explicitly declares a variable/symbol as `Tuple<...>` I feel like the `ItemN` properties should become available (while otherwise still behave exactly the same).
1. This would solve this bug as we would gain backwards compatibility
2. It removes the surprise of people going to the [docs](https://msdn.microsoft.com/en-us/library/dd268536(v=vs.110).aspx) and expecting the `ItemN` properties.

Besides that I feel like `X*Y` should behave like `Tuple<X,Y>`. Is there a problem with this suggestion? Is it too hard to implement? Basically the compiler has to "remember" if the type was declared as F# tuple or `System.Tuple` and unify them later (if at all).

I think what I'm missing in the above solution is how we can differentiate F# tuples from `System.Tuple<...>` in external references. Is it even feasible to detect if a function type was declared as F# or `System.Tuple` after compilation? Probably not. But then I'd favor @dsyme's argument of not enabling `ItemN` everywhere.

**Comment by dsyme** [comment]
> Is there a problem with this suggestion? Is it too hard to implement? Basically the compiler has to "remember" if the type was declared as F# tuple or System.Tuple and unify them later (if at all).

@matthid Realistically we could only do that if they were considered distinct nominal types (F# makes an effort to remember abbreviations but these are not necessarily preserved by type inference).  

Note F# has always translated ``Tuple<X,Y>`` encountered in .NET libraries to ``(X * Y)``.   The thing we were not doing previously was translating ``Tuple<X,Y>`` declared in **F#** code to ``(X * Y)``.

**Comment by curtnichols** [comment]
Here's my latest:

The return of this `Func` is is what broke:

```
type FileMapper = Func<string, System.Tuple<bool, Stream>> // Broken in VS 15.4 where .ItemN is used.
```

This version using F# idioms would be _very bad_ interop:

```
type FileMapper = string -> (bool * Stream) // Bad interop: FSharpFunc<string, Tuple<bool, Stream>> MapFilePath
```

And this works on both 15.3.5 and 15.4. Both versions of VS generate a `System.Tuple` as the return value.

```
type FileMapper = Func<string, (bool * Stream)> //  Func<string, Tuple<bool, Stream>> MapFilePath
```

Note that spelling the tuple in the F# dialect left me with the `System.Tuple`, which is what I want for intrerop. At the moment I'm just not sure why. :)

**Comment by dsyme** [comment]
> Note that spelling the tuple in the F# dialect left me with the System.Tuple, which is what I want for intrerop. At the moment I'm just not sure why it did what I want. :)

In F# ``A * B`` and ``Tuple<A,B>`` are now two syntaxes for exactly the same thing.  They are identical.

It's just that F# doesn't let you use the ItemN properties on this thing anymore, for reasons mentioned above.  

And yes, we should not have made that breaking change except via an RFC, in which case I suppose we would surely have supported Item1, ... at least for small tuples.

**Comment by KevinRansom** [comment]
@dsyme what do you propose for next steps on this PR?

Thanks

Kevin

**Comment by dsyme** [comment]
@KevinRansom I believe we are at the point where we will just have to accept that this breaking change happened.

**Comment by cartermp** [comment]
Part of the problem here is we have no data on how often people snag `ItemX` from a `System.Tuple` defined in F# code.  I suspect that few people do this in a way that is difficult to correct, but the mere fact that there is a breaking change here will keep people and orgs from adopting VS 2017, even if they're not broken by this at all.  I'm inclined to say this should be fixed, even though it requires a change to the language to allow `Item1`-`Item7` access on F# tuples.  We'll just have to _not_ document that so that it doesn't catch on.

**Comment by dsyme** [comment]
@cartermp OK, I'll think over the cost and approach to factor that in  

> We'll just have to not document that so that it doesn't catch on.

It's not that I think ``ItemN`` properties are bad - they actually fill a need since utility functions such as now ``let p13 (a,b,c) = c`` aren't really needed, especially if ``_.Item1`` becomes allowed as a first class function value (see RFC on that).   It would be wrong to steop them appearing in intellisense for example, if they are legitimate properties

It's just it's best discussed

**Comment by cartermp** [comment]
Agreed

**Comment by cartermp** [comment]
@KevinRansom what is `UpdateArgs`?  In this case `result` can be converted directly into `success, file`:

```fsharp
[<CompiledName("F")>]
let f (boolStream : Tuple<bool, System.IO.Stream>) =
    let success, file = boolStream
    printfn "%A %A" success file.CanRead
```
Why does `ItemX` property access have to work aside from strictly from a "not breaking existing code" reason?

**Comment by KevinRansom** [comment]
@cartermp 
The developer specified that result is a Tuple<bool, Stream>.  In every other .Net language it would have an Item1 and Item2 property.  It should in F# too.  That F# has a nicer deconstruction syntax, is irrelevant.

Kevin

**Comment by dsyme** [comment]
> In every other .Net language it would have an Item1 and Item2 property. It should in F# too. That F# has a nicer deconstruction syntax, is irrelevant.

If life were only made up of small tuples I'm fine with this approach.

However, for large tuples size > 7), other .NET languages see the compiled representation of tuples. Normal F# code suppresses the compiled representation from view. From F#, the compiled representation can be observed in multiple ways (reflection, typeof etc) but in normal F# coding the user is not aware of that encoding. 

For example, what would Item8 show for an 9-tuple of integers?  A pair of integers??  (That's the encoding used for large tuples in the compiled representation, and I presume that's what C# shows). Or what intellisense list do you get when you press ``.`` for for a 100-tuple?  ``Item1`` to ``Item8`, where ``Item8`` has a massive whacky type?   

It's a reasonable choice to make the ``Item`` properties available for small tuples - especially for backwards compat reasons - but I'm loathe to do it for large tuples. But equally, if we don't make them available for large tuples, then programmers can get to a situation where adding an extra element to a tuple suddenly means all those ``Item`` properties are no longer available. 

A separate consideration is that using ``Item`` properties for tuple elements arguably makes code less readable - when you are forced to use tuple decomposition with a pattern you are forced to give names to the elements of the tuple, which is valuable metadata, c.f. 

    (fun p -> p,Item1 / p.Item2) 

v. 

    (fun (distance, time) -> distance / time)

or even

    (fun (d,t) -> d/t)

Both the second and third carry more useful metadta - the first is horrible for readability and maintainability. People are lazy and love hitting ``.``. But I don't think it will remotely increase the quality of F# code being written.  

(As an aside, note that, after the change that caused this regression, the F# type system can't effectively distinguish between ``int * int`` and ``Tuple<int,int>`` - a choice you make about one form of declaring tuple types applies to the other.)

**Comment by matthid** [comment]
@dsyme I think we make our life harder than it actually is:

We probably should not even try to change regular F# code for this bugfix. The Compiler should introduce/allow `ItemX` properties only when the type is annotated explicitly as .NET `Tuple` type. This is what a programmer expects. That `Tuple<X,Y>` and `X*Y` is the same is an implementation detail (from a language point of view they can be converted implicitly from one to another)   

> Note F# has always translated Tuple<X,Y> encountered in .NET libraries to (X * Y). The thing we were not doing previously was translating Tuple<X,Y> declared in F# code to (X * Y).

So all we need to do to fix this bug is allow `ItemX` in that particular case. We don't introduce any problems regarding bigger tuples as it is hard to explicitly annotate them anyway (they don't actually exist as explicit `Tuple` types).

Maybe we shouldn't try to add a general language feature for tuples around here and somehow concentrate on the one special case we broke?

So in short: Can't we have a solution without `ItemX` in the general case (we might be able to add that later) and still recover this edge case? It probably is a bit of work...

Above I was thinking that it is hard to do, because of external libraries. But as we never supported that case anyway we don't have support that case here in the bugfix?


Leaving this bug aside. It really feels strange to explicitly annotate something with a regular .NET type and then not having access to its members (as @KevinRansom has already said). Do we have something similar somewhere else in the language?

**Comment by mexx** [comment]
IMHO there are two ways of approaching the problem.

On one side, if we see F# tuple and System.Tuple as synonyms, the API provided by either of the types should contain all of the possible properties and methods defined on any of them. The types should be made really interchangeable. I think this was as well the rationale behind the changes whose leaded to the current state.

On the other side, we can treat F# tuple and System.Tuple as implementation equal but different types on the type system with an implicit conversion between them as they are equal on implementation. When the type is explicitly provided it shouldn't change in the type system, but the value should be automagically able to be used as the other "implementation equal" type. This approach would make the types as well interchangeable.

For me the second approach would fix the problem for the interop scenario (as @matthid described) and as well would allow to proceed with the first approach once this identified to be needed for example for the `_.ItemX` syntax mentioned by @dsyme.

**Comment by dsyme** [comment]
Re synonyms v. distinct.  We have long regarded  F# tuple and System.Tuple as _synonyms_ when 
1. ``System.Tuple`` appears in the metadata of a non-F#  .NET assembly, and
2. The particular use of System.Tuple conforms to a valid compiled representation of an F# tuple type.

Prior to the breaking change we regarded them as _distinct_ when they appeared explicitly in F# source code.  However that approach led to unresolvable logical inconsistencies in corner cases.

Clearly we need to continue to regard them as synonyms in the cases where we already do that, we have no choice. 

> Compiler should introduce/allow ItemX properties only when the type is annotated explicitly as .NET Tuple type. This is what a programmer expects. 

While I sympathize with this, especially since it was the behaviour prior to the change, I don't necessarily want to follow "what the programmer expects" here.

1. In the F# design and implementation it is extremely difficult to attempt to allow two types that are "synonyms" to be non-equivalent, e.g. for the purposes of name resolution of properties.   And I mean really, really difficult, among other things it results in incoherent, under-specified or fragile behaviour in type inference.  

2. I'm not particularly happy to have the ``Item`` properties appearing in intellisense lists just because some C# programmer rocks up and complains "where are my Item properties".  Tuples are very fundamental to F# - we shouldn't muck with their usability/engineering/resulting-code-quality properties just because of that.   As I said above, if we do add the Item properties to intellisense lists (for F# tuple values) they will get used _all_ _the_ _time_.  I know the proposal is to only add them for ``TUple<_,_>`` annotated types, but that's very hard, see point 2 above

3. Obviously for large tuples it is fundamentally "wrong" for ``Item8`` to reveal the nested encoding (which we would be at risk of given that we basically can't have one view of the properties for ``( ... * ... * ...)``-declared or inferred large tuple types, and one for ``Tuple<...>`` declared tuple types)

Note also that large tuple types are common enough in F#, because they are automatically inferred for large tuple expressions.

Given all this, a design that would satisfy me is:
1. Allow `Item*`` properties to be name-resolved on all small tuple types (size <= 7), and
2. Hide all ``Item*`` properties from intellisense lists, and
3. Give a warning that their use should be replaced by ``let (x,y) = p`` tuple decomposition in F#.  

That's basically where we want code to end up, and would deal with the breaking change.

As an aside, there are precedents for hiding things from intellisense lists in F# where we don't want them to infest F# code and thinking, e.g. we don't show ``Equals`` and ``GetHashCode`` for F# types that don't satisfy the equality constraint, see https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/NameResolution.fs#L4141.

**Comment by curtnichols** [comment]
@dsyme I like your solution, Don. It un-breaks broken code and discourages the `.ItemX` usage. The new warning would be treated as an error in our build, but I presume it would come with a descriptive warning message. We dealt with the one line of `.ItemX` in our F# code base and moved on.

As a C/C++/C#/F# (primarily ;)), I have long found the `.ItemX` usage in C# to be inelegant. As far as C#-flavored developers being concerned about lack of `.ItemX` in F#:

1. the new warning would help them write more idiomatic F# code, and
1. C# now has better tuple support, meaning they don't need to use the `.ItemX` syntax in C#, either. That's a welcome improvement in my view.

**Comment by ReedCopsey** [comment]
Just a small comment:

> For example, what would Item8 show for an 9-tuple of integers? A pair of integers?? (That's the encoding used for large tuples in the compiled representation, and I presume that's what C# shows). Or what intellisense list do you get when you press . for for a 100-tuple? Item1 to Item8, where Item8 has a massive whacky type? That's just wrong.

It's very ugly in C# - you need to understand the representation (which is crazy), and do `var item9 = theTuple.Rest.Item1.Item2;`.  This is because "Rest" returns the `Tuple<Rest>` (1 item tuple) containing the tuple of the 8th and 9th values internally.

As `Item8` does not exist, I'm not sure this is necessary:

> Allow `Item*`` properties to be name-resolved on all small tuple types (size <= 7), and

Allowing the `Rest` property to be name resolved (but hidden from intellisense) doesn't seem inappropriate if you're allowing the first 7 items _and_ including a warning already. This would prevent the above ugliness from being a breaking change, but the warning/hidden intellisense would still be enough to discourage people from using that monstrosity.

**Comment by gusty** [comment]
I agree to add for completeness reasons the property ``Rest`` no matter what crazy encoding it reveals. I don't think even in C# people goes into property Rest frequently, it is there for completeness and should be added here as well for the same reasons.

To me it will look even crazier to provide access the internals of a 7-uple but not for the 8-uple.

**Comment by KevinRansom** [comment]
@dsyme 

I think the proposal is:
•	For Tuples and ValueTuples allow Items 1-7 + Rest.
•	Warn on use with improved F# syntax …. (Off by default since it’s a breaking change?)
•	Hide Tuple members from intellisense?

What should intellisense do with this:
let x = struct(1, 2)
printfn “%d” x.

**Comment by dsyme** [comment]
> For Tuples and ValueTuples allow Items 1-7 + Rest.
> Warn on use with improved F# syntax …. (Off by default since it’s a breaking change?)
> Hide Tuple members from intellisense?

Yes, though 
* I don't care about Rest, I don't believe anyone has used it
* historically haven't considered adding "sensible" warnings as true breaking changes

That said, the above is a fair chunk of work and the issue doesn't appear to be of screeching urgency.

**Comment by gusty** [comment]
> I don't care about Rest, I don't believe anyone has used it

I have used Rest. It's not a common pattern but I think we should definitely have it available, again for completeness. I like the way F# handle tuples, in the sense that it's not theoretically limited to a certain number of elements, as it is in other languages, so let's keep it like that.

**Comment by dsyme** [comment]
> I have used Rest.  It's not a common pattern but I think we should definitely have it available, again for completeness. 

You've used Rest from F# code?  Why?

We are treating this as a tactical fix to undo a breaking change - but not to encourage or extend how F# programmers program with tuples. Given that the cost of this fix is fairly high and we've little evidence of any widespread use of ``Item*`` properties on explicit F# uses of ``System.Tuple<...>`` I'm tempted just to not do this at all.  We've no evidence at all of the use of ``Rest``....

**Comment by gusty** [comment]
> You've used Rest from F# code? Why?

I don't remember exactly the reasons, but I think it was to generalize at some point in the code over arbitrary size tuples and avoid a lot of boilerplate while at the same time have strong type checking.
That part of the code was something like a little dataframe lib, but strongly typed.
Since I have no longer access to the source code I can't give you more details, but I know they're using it in production.

If it's a big effort I'm happy to help on that part.

**Comment by ingted** [comment]
Hi,

I usually use FsPickler and when I change the code like this:

https://github.com/mbraceproject/FsPickler/pull/97#issue-275127733

And I could build it successfully on my Windows Server 2016 with clear VS 2017 update 15.4.
I did a git commit and try to pull request, but the CI tool just failed to compile it.

https://travis-ci.org/mbraceproject/FsPickler/builds/304174513?utm_source=github_status&utm_medium=notification

This is good for those who usually follow the latest version tool and quickly change their code (e.g. me?!) but not comfortable with the others who only NuGet something to use...

**Comment by eiriktsarpalis** [comment]
@ingted @dsyme +1, this regression makes it impossible to build FsPickler without resorting to reflection.

**Comment by gusty** [comment]
> @ingted @dsyme +1, this regression makes it impossible to build FsPickler without resorting to reflection.

... and it worth clarifying that ``.Rest`` is also [being used in FsPickler](https://github.com/mbraceproject/FsPickler/search?utf8=%E2%9C%93&q=Rest&type=).

**Comment by dsyme** [comment]
> ... and it worth clarifying that .Rest is also being used in FsPickler.

Yeah, I saw that.  

@cartermp One approach to this breaking change would simply be to revert https://github.com/Microsoft/visualfsharp/pull/3283 and try again with that change in F# 4.2.

**Comment by cartermp** [comment]
@dsyme I agree that reverting #3283 is a way to go, but such a change will not be until at least the VS 2017 15.6 release.  We're currently in Ask Mode for the first Preview (group engineering manager approval required for changes), but the final release is not likely to be for a few months.  There is a chunk of time allocated for addressing issues with the large forthcoming 15.5 release.  And we're far, far beyond the time where we could get a compiler change into 15.5.

We could also look into shipping a VSIX through the gallery, but this will not have the same reach as VSIX distributed in-box, plus it's another thing we'll need to account for until 15.6 RTW.  So I think we should look for a way to unblock FsPickler in the interim.

I'm a bit confused about the issue with FsPickler.  Looking at the code, it should work with the new compiler with small tweaks.  @dsyme says this:

> In F# `A * B` and `Tuple<A,B>` are now two syntaxes for exactly the same thing. They are identical.

However, in a [PR intended to unblock FsPicker](https://github.com/mbraceproject/FsPickler/pull/97), @eiriktsarpalis says this:

> Thanks, but Tuple`8 types are not exactly the same as a tuple of 8 elements. In order to get this to work, the F# compiler issue must be fixed.

Does the FsPickler build contain the same compiler as #3283?  If not, can we try to update it and see what happens?

**Comment by dsyme** [comment]
Yes, I understand it would be for 15. at the earliest.

> I'm a bit confused about the issue with FsPickler. Looking at the code, it should work with the new compiler with small tweaks.

My understanding is that FsPickler programs over ``Tuple`8`` (i.e. ``Tuple<t1,...,t8>``) with explicit access to ``Rest`` in order to deal with arbitrarily sized tuples without using reflection.  That programming technique is not possible after #3283 (partly because accessing ``Rest`` is not possible, and particular because an F# 8-tuple ``(int * .... * int)`` corresponds only tuples of logical size 8, where ``:? Tuple`8`` matches tuples of logical size 8 and greater).

So @eiriktsarpalis is correct that it's not possible after #3283.

**Comment by cartermp** [comment]
So that means that [this code](https://github.com/mbraceproject/FsPickler/pull/97/files#diff-13f25f5240116cff5e674543eaa0057fR323) basically won't work?  Suppose I could just pull it down and try it out.

**Comment by dsyme** [comment]
> So that means that this code basically won't work? Suppose I could just pull it down and try it out.

Correct - specifically it won't work on a 9+ tuple

**Comment by cartermp** [comment]
Okay, in that case let's revert and try to get it into the 15.6 preview

**Comment by eiriktsarpalis** [comment]
FWIW here's a small snippet that illustrates how tuples of large arities are encoded:
```fsharp
open FSharp.Reflection

let mkTupleTy arity = 
    Array.init arity (fun _ -> typeof<int>)
    |> FSharpType.MakeTupleType
    |> string

Array.init 20 ((+) 1 >> mkTupleTy)
```
8-tuples are cannot work either, because they also rely on nesting a 1-tuple in the `Rest` parameter. In general it is not possible to pattern match on a `Rest` argument using regular tuple expressions.

**Comment by dsyme** [comment]
> Okay, in that case let's revert and try to get it into the 15.6 preview

@cartermp See #4029, though as I note in the PR it makes me feel queasy.  The alternative is to implement the proposal above after all:

•	For Tuples and ValueTuples allow Items 1-7 + Rest.
•	Warn on explicit use of Item* and Rest
•	Hide Tuple members from intellisense

**Comment by 0x53A** [comment]
> Warn on explicit use of Item* and Rest

Would that really be necessary in your opinion?

I find destructuring larger tuples to be annoying (even though large tuples are probably an anti-pattern).

Compare 

```
let (_,__,x,_,_,_) = tpl
let x = tpl.Item3
```

In the first case it is really not obvious which member is accessed - I need to count the number of ignored items.

This issue is compounded by the fact that it is not possible to write universal accessor functions ``fst``, ``snd``, or ``itemN`` which work over all tuple sizes. So I would need to create ``fstOf4``, ``fstOf5`` and so on.

**Comment by 0x53A** [comment]
In the current state, it seems also impossible to explicitly call the constructor of System.Tuple:

```F#
let tpl3 = new System.Tuple<int,int,int>(8,9,10)
```

> error FS0756: 'new' may only be used with named types

**Comment by dsyme** [comment]
@0x53A Re ``Item``

I understand the temptation to add these to the F# programming model for all tuple values.  However, see my comment above, particularly:

> A separate consideration which is weighing on my mind is that using Item properties for tuple elements makes code less readable... When you are forced to use tuple decomposition with a pattern you are forced to give names to the elements of the tuple, which is valuable metadata,.....[using Item properties] is horrible for readability and maintainability. People are lazy and love hitting ``.`` ....making Item properties available in intellisense lists will not remotely increase the quality of F# code being written, and indeed will harm it.

**Comment by dsyme** [comment]
> In the current state, it seems also impossible to explicitly call the constructor of System.Tuple:

Yes, that is correct.  The proposal [here](https://github.com/Microsoft/visualfsharp/pull/3729#issuecomment-347590956) would be not to change this.

**Comment by 0x53A** [comment]
> [...] When you are forced to use tuple decomposition with a pattern you are forced to give names to the elements of the tuple, which is valuable metadata,..... [...]

This is a nice goal, but I don't think it actually holds in practice. In reality I will use

``let (_,__,x,_,_,_) = tpl`` 

where x doesn't really tell me anything, and I lose the information which member I access and so need to manually count.

----

IF you really hate ItemX, please consider https://github.com/fsharp/fslang-suggestions/issues/628. The SRTP method needs ItemX visible, but it could still warn on explicit use, and/or hide them from intellisense, as long as they are visible to the constraint solver.

**Comment by dsyme** [comment]
> In reality I will use ``let (_,__,x,_,_,_) = tpl`` where x doesn't really tell me anything, and I lose the information which member I access and so need to manually count.

I agree this kind of code is poor and in good, long-lived code there should be accurate transition paths away from it. Using a large tuple or a name ``x`` that is a programmer's problem.  However, there are progressive, accurate steps the programmer can take to rectify the mess, e.g. turn the tuple values into a record type or a single-case discriminated union.

In contrast offering ``tpl.Item3`` in the intellisense list is worse 
1. We actively incentivize the use of a crappy, meaningless name (even ``x`` is better than ``Item3``)
2. There is no incentive to move away from a large tuple 
3. Programming is less accurate under change. For example, consider when an extra item is added or removed from the front of the tuple - you have to manually hunt down and correct all ``Item`` accesses that aren't caught by the type checker.  With an explicit tuple, you at least manually go and adjust the tuple patterns that consume the tuple.

**Comment by 0x53A** [comment]
Do you think the same about external accessor functions (fst, snd, thrd, ...)?

The big advantage of them is that they aren't easily discoverable (you can't dot into them from the tuple itself), so you can only use them if you already know they are there ;-)

**Comment by dsyme** [comment]
> Do you think the same about external accessor functions (fst, snd, thrd, ...)?

I'm not fundamentally opposed to generic constructs over generic data shapes.  For tuple projections I think the incentives are set about right for those in F# today:

* ``fst`` and ``snd`` are there but you aren't particularly incentivised to use them

* You can define additional projections (e.g. ``p13``, ``p14`` as we do in the compiler code) but you probably know what you're doing in that case

**Comment by dsyme** [comment]
The fix for this issue is at https://github.com/Microsoft/visualfsharp/pull/4034

**Comment by dsyme** [comment]
The RFC issue for this whole tuple malarkey is at https://github.com/fsharp/fslang-design/blob/master/FSharp-4.1b/FS-1046-consistent-tuple-types.md

**Comment by KevinRansom** [comment]
Thanks for this,

I am closing because I believe this is handled by Don's PR: https://github.com/Microsoft/visualfsharp/pull/4034
